src/persist-client/src/internal/apply.rs

```diff
@@ -83,68 +88,145 @@ where
+        self.upper()
+    }
+
+    /// A point-in-time read of `upper` from the current state.
+    ///
+    /// Due to sharing state with other handles, successive reads to this fn or any other may
+    /// see a different version of state, even if this Applier has not explicitly fetched and
+    /// updated to the latest state.
+    pub fn upper(&self) -> Antichain<T> {
+        self.state
+            .read_lock(&self.metrics.locks.applier_read_cacheable, |state| {
+                state.upper().clone()
+            })
+    }
+
+    /// A point-in-time read of `since` from the current state.
+    ///
+    /// Due to sharing state with other handles, successive reads to this fn or any other may
+    /// see a different version of state, even if this Applier has not explicitly fetched and
+    /// updated to the latest state.
+    #[cfg(test)]
+    pub fn since(&self) -> Antichain<T> {
+        self.state
+            .read_lock(&self.metrics.locks.applier_read_cacheable, |state| {
+                state.since().clone()
+            })
+    }
+
+    /// A point-in-time read of `seqno` from the current state.
+    ///
+    /// Due to sharing state with other handles, successive reads to this fn or any other may
+    /// see a different version of state, even if this Applier has not explicitly fetched and
+    /// updated to the latest state.
+    pub fn seqno(&self) -> SeqNo {
+        self.state
+            .read_lock(&self.metrics.locks.applier_read_cacheable, |state| {
+                state.seqno
+            })
+    }
+
+    /// A point-in-time read of `is_tombstone` from the current state.
+    ///
+    /// Due to sharing state with other handles, successive reads to this fn or any other may
+    /// see a different version of state, even if this Applier has not explicitly fetched and
+    /// updated to the latest state.
+    pub fn is_tombstone(&self) -> bool {
+        self.state
+            .read_lock(&self.metrics.locks.applier_read_cacheable, |state| {
+                state.collections.is_tombstone()
+            })
+    }
+
+    /// A point-in-time read of `seqno_since` from the current state.
+    ///
+    /// Due to sharing state with other handles, successive reads to this fn or any other may
+    /// see a different version of state, even if this Applier has not explicitly fetched and
+    /// updated to the latest state.
+    pub fn seqno_since(&self) -> SeqNo {
+        self.state
+            .read_lock(&self.metrics.locks.applier_read_cacheable, |state| {
+                state.seqno_since()
+            })
+    }
+
+    /// A point-in-time read of `since` and `upper` from the current state.
+    ///
+    /// Due to sharing state with other handles, successive reads to this fn or any other may
+    /// see a different version of state, even if this Applier has not explicitly fetched and
+    /// updated to the latest state.
+    pub fn since_and_upper(&self) -> (Antichain<T>, Antichain<T>) {
```

Comment:
this is only used in one place which just calls `!since.is_empty() || !upper.is_empty()` on them. let's move the `is_empty` checks in here and rename it (perhaps `since.is_empty() && upper.is_empty()` and then name it `since_upper_both_empty`? then `!since_upper_both_empty()` is equivalent to the original logic)

(we could also remove the method entirely, I think. AFAICT the reads don't need to be from the same version of state, but may as well keep the unnecessary behavior changes down in this PR and also it's nice to only grab the lock once)